### ISSUE TITLE
Clusterctl upgrade does not respect config for n-Infra setup

### ISSUE THREAD

**Issue by MaxRink** [kubernetes-sigs/cluster-api#2997]
**What steps did you take and what happened:**
We are running an n-Infra Setup with different credentials for the infra providers

```
clusterctl --kubeconfig mgmt-capi-cluster-1.kubeconfig upgrade plan
Checking new release availability...

Management group: capi-system/cluster-api, latest release available for the v1alpha3 API Version of Cluster API (contract):

NAME                     NAMESPACE                           TYPE                     CURRENT VERSION   NEXT VERSION
bootstrap-kubeadm        capi-kubeadm-bootstrap-system       BootstrapProvider        v0.3.5            Already up to date
control-plane-kubeadm    capi-kubeadm-control-plane-system   ControlPlaneProvider     v0.3.5            Already up to date
cluster-api              capi-system                         CoreProvider             v0.3.5            Already up to date
infrastructure-vsphere   capv-system-1                 InfrastructureProvider   v0.6.4            Already up to date
infrastructure-vsphere   capv-system-2                InfrastructureProvider   v0.6.4            Already up to date

You are already up to date!
```

Im im doing a clusterctl upgrade the credentials used by the 2nd instance get overwritten with those of the 1st instance.
Thus i have to patch the 2nd secret after the update and kill the 2nd pod.
**What did you expect to happen:**
either do an upgrade in steps, so i can specify different configs with the different credentails needed or having support for credential matching of providers within one config.

**Anything else you would like to add:**

**Environment:**

- Cluster-api version: 0.3.5

/kind bug

**Comment by vincepri** [comment]
@MaxRink Just to clarify a little more, are you seeing that the second infrastructure provider's credential (in your case the one in `capv-system-2` namespace) credentials are being lost/overwritten by the ones in the other namespace?

**Comment by vincepri** [comment]
/assign @wfernandes @fabriziopandini 

to investigate

/priority critical-urgent
/milestone v0.3.6

**Comment by ncdc** [comment]
@MaxRink can you share the owner references for both secrets prior to running the upgrade?

**Comment by fabriziopandini** [comment]
@MaxRink 

So the key point here is that AWS components has some variable and you want to keep this variable  instance-specific.

When creating the management cluster I assume you achieved this running `clusterctl init` several times, each one with different env variables.

In order to preserve this behavior, you should do the same during upgrades. This means that you you have to upgrade in steps by calling
- `clusterctl upgrade apply --core --bootstrap --control-plane` (with the specific upgrade target for each one as printed by upgrade plan) once
- `clusterctl upgrade apply --infrastructure` for each infrastructure provider instance / with different env variables

Please note that this works for upgrades within the same Cluster API contract version (v1alpha3->v1alpha3) only;
@ncdc @vincepri should we start designing a potential solution for a similar problem in case of v1alpha3->v1alpha4 upgrades?

**Comment by vincepri** [comment]
/kind documentation
/milestone v0.3.9

We won't fix this in v0.4.x given that we're planning on supporting multi-tenancy clusters with a single controller deployment.
